smolyay/samples.py

```diff
@@ -0,0 +1,496 @@
+    exponentially with increasing n.
+    
+    While the total number of unique points increases exponentially with
+    increasing n, the rate of increasing points with respect to level L can be
+    limited at or below a linear :math:`2*L + 1`. To do this, if adding the
+    next set of points at a level L causes the rate of new points to go above
+    :math:`2*L + 1`, then the points will not be added and the level will be
+    empty.
+
+    The order :math:`o(L) of the slow nested Clenshaw Curtis set describes
+    the number of cummulative points at each level to account for the limiting
+    rate :math:`2*L + 1`. Then is used to determine the number of points at
+    each individual level. The order of the slow Clenshaw Curtis set is
+    described by the following equation:
+
+    .. math::
+
+        o(L) = \begin{cases}
+                1 & \text{ if } L = 0\\ 
+                2^{k} + 1 & \text{ if } L > 0 
+        \end{cases}
+
+        where k = \left \lceil \log_{2}(L) \right \rceil + 1
+
+    The sequence of o(L) is then :math:`{1, 3, 5, 9, 9, 17, ...}`.
+    
+    Since the points are nested, the number of unique points at
+    a level L is 
+    
+    .. math::
+        num_per_level(L) = o(L) - o(L - 1)    
+
+    The points are then scaled from the domain :math:`[-1, 1]` to the domain
+    specified by the :attr:`domain`.
+
+    Parameters
+    ----------
+    domain: list
+        Domain of the sample points.
+
+    num_level: int
+        The number of levels.
+    """
+
+    def _create(self):
+        r"""Create the points in the set.
+
+        Generating nested extrema of chebyshev polynomials of the first kind.
+        """
+        # create properties for levels, level 0 is a special case with 1 point
+        rule = lambda x: 1 if x == 0 else int(2 ** (numpy.ceil(numpy.log2(x)) + 1) + 1)
+        self._num_per_level = numpy.ones(self.num_level, dtype=int)
+        self._num_per_level[1:] = [
+            rule(i) - rule((i - 1)) for i in range(1, self.num_level)
+        ]
+        self._end_level = numpy.cumsum(self._num_per_level)
+        self._start_level = self._end_level - self._num_per_level
+
+        # points, level 0 is a special case only 0 as a point
+        num_points = self._end_level[-1]
+        points = numpy.zeros(num_points, dtype=float)
+        for i in range(1, self.num_level):
+            if self._num_per_level[i] == 0:
+                continue
+            # find indexes of extrema not already found. Fraction index/degree
+            # cannot be further simplified
+            degree = int(2 ** (numpy.ceil(numpy.log2(i)) + 1))
+            if i == 1:
+                # special case for level == 1
+                indexes = numpy.arange(0, degree + 1, 2, dtype=int)
+            else:
+                indexes = indexes = numpy.arange(1, degree, 2, dtype=int)
+                indexes = indexes[~(numpy.gcd(indexes, degree) > 1)]
```

Comment:
Please clean up this code to make it more readable like we did for the other nested CC class.